### PR TITLE
Reapply the OptimizedModule in Fabric.setup()

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #17250

When you do

```py
model = torch.compile(orig_model)
model = fabric.setup(model)
```

today, then you end up with
```py
FabricModule(StrategyWrapper(OptimizedModule(orig_model)))
```
where StrategyWrapper represents the wrapper applied by the strategy, e.g., FSDP, DDP, DeepSpeedEngine, etc.

With the new feature enabled in this PR, you will now get:

```py
FabricModule(OptimizedModule(StrategyWrapper(orig_model)))
```
Note how the optimized module gets applied outside the strategy wrapper, but inside the fabric module router.
It is important to have the fabric module outside the optimized module, because it contains only convenience routing and precision management that does not need to be compiled (nor can it be without graph breaks atm). At the same time, the FSDP/DDP specific operations can still be captured properly by torch compile, and how it is intended to be wrapped in PyTorch.

This feature is marked as **experimental** because we first need to test its application in the wild.

Follow-up work:
- Verify it in the wild
- Decide on the default
- Document it in the compile-guide
- Do the same in Lightning Trainer

----
📚 Documentation preview 📚: https://pytorch-lightning--19192.org.readthedocs.build/en/19192/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/fabric.py [hunk 1]
```diff
@@ -213,6 +214,7 @@ def setup(
         module: nn.Module,
         *optimizers: Optimizer,
         move_to_device: bool = True,
+        _reapply_compile: Optional[bool] = None,
```
Comment (Borda) [inline src/lightning/fabric/fabric.py:217]: Interesting, I never saw a private argument...

**Review by Borda — APPROVED** [review]

File: src/lightning/fabric/fabric.py [hunk 2]
```diff
@@ -221,13 +223,18 @@ def setup(
...
+                same settings after the model was set up by the strategy (e.g., after the model was wrapped by DDP,
+                FSDP etc.). Only supported on PyTorch >= 2.1. Defaults to ``False``, but it may change in the future.
 
         Returns:
             The tuple containing wrapped module and the optimizers, in the same order they were passed in.
 
         """
         self._validate_setup(module, optimizers)
         original_module = module
+        module, dynamo_context = _unwrap_compiled(module) if _reapply_compile else (module, None)
```
Comment (carmocca) [inline src/lightning/fabric/fabric.py:237]: nit: I would prefer a more generic name here

```suggestion
        module, compile_data = _unwrap_compiled(module) if _reapply_compile else (module, None)
```

File: src/lightning/fabric/fabric.py [hunk 3]
```diff
@@ -221,13 +223,18 @@ def setup(
...
+                corresponding :class:`~torch._dynamo.OptimizedModule` wrapper will be removed and reapplied with the
+                same settings after the model was set up by the strategy (e.g., after the model was wrapped by DDP,
+                FSDP etc.). Only supported on PyTorch >= 2.1. Defaults to ``False``, but it may change in the future.
 
         Returns:
             The tuple containing wrapped module and the optimizers, in the same order they were passed in.
 
         """
         self._validate_setup(module, optimizers)
         original_module = module
```
Comment (carmocca) [inline src/lightning/fabric/fabric.py:236]: Should we unwrap before saving a reference to the original module? That could be useful.

File: src/lightning/fabric/fabric.py [hunk 4]
```diff
@@ -242,6 +249,8 @@ def setup(
         else:
             module = self._strategy.setup_module(module)
 
+        if dynamo_context is not None:
+            module = _to_compiled(module, dynamo_context)
         module = _FabricModule(module, self._precision, original_module=original_module)
```
Comment (carmocca) [inline src/lightning/fabric/fabric.py:254]: Very good. This pattern adheres to https://github.com/Lightning-AI/pytorch-lightning/issues/17250#issuecomment-1831013511

File: src/lightning/fabric/fabric.py [hunk 5]
```diff
@@ -410,6 +429,7 @@ def backward(self, tensor: Tensor, *args: Any, model: Optional[_FabricModule] =
 
         """
         module = model._forward_module if model is not None else model
+        module, _ = _unwrap_compiled(module)
```
Comment (carmocca) [inline src/lightning/fabric/fabric.py:432]: Why do you need to unwrap here and in `no_backward_sync`?

File: tests/tests_fabric/strategies/test_fsdp_integration.py [hunk 6]
```diff
@@ -373,6 +374,30 @@ def test_compile(compile_after_setup):
...
+@mock.patch.dict(os.environ, {})
+def test_reapply_compile():
+    """Test that Fabric can rewrap a compiled module such that compilation happens over the FSDP-wrapper."""
+    from torch._dynamo import OptimizedModule
+
+    fabric = Fabric(accelerator="cuda", devices=2, strategy="fsdp")
+    fabric.launch()
+
+    model = BoringModel()
+    compiled_model = torch.compile(model, mode="reduce-overhead")
```
Comment (carmocca) [inline tests/tests_fabric/strategies/test_fsdp_integration.py:387]: If this has no graph breaks, then you can do:

```suggestion
    compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
```

File: tests/tests_fabric/strategies/test_fsdp_integration.py [hunk 7]
```diff
@@ -344,6 +344,7 @@ def test_setup_with_orig_params_and_multiple_param_groups():
         assert not isinstance(layer.weight, FlatParameter)
 
 
+# TODO: Do we still need a separate test or should we combine it?
```
Comment (carmocca) [inline tests/tests_fabric/strategies/test_fsdp_integration.py:347]: Combining makes sense to me

File: src/lightning/fabric/fabric.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/fabric/fabric.py:236]: Ok done, addressed in #19280

File: src/lightning/fabric/fabric.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning/fabric/fabric.py:432]: For completeness, it's optional.

File: tests/tests_fabric/strategies/test_fsdp_integration.py (hunk 6 above)
Comment (awaelchli) [reply tests/tests_fabric/strategies/test_fsdp_integration.py:387]: Unfortunately it does graph-break.
```
        # Smoke-testing forward to ensure we don't get compilation errors
        for _ in range(3):
>           fabric_model(torch.randn(2, 32, device=fabric.device))

tests/tests_fabric/strategies/test_fsdp_integration.py:407: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/nn/modules/module.py:1511: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/nn/modules/module.py:1520: in _call_impl
    return forward_call(*args, **kwargs)
src/lightning/fabric/wrappers.py:138: in forward
    output = self._forward_module(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/nn/modules/module.py:1511: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/nn/modules/module.py:1520: in _call_impl
    return forward_call(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py:417: in _fn
    return fn(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py:580: in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:384: in _convert_frame_assert
    return _compile(
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:643: in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/utils.py:247: in time_wrapper
    r = func(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:524: in compile_inner
    out_code = transform_code_object(code, transform)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py:1033: in transform_code_object
    transformations(instructions, code_options)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:151: in _fn
    return fn(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:489: in transform
    tracer.run()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2098: in run
    super().run()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:780: in run
    and self.step()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:743: in step
    getattr(self, inst.opname)(inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:462: in wrapper
    return inner_fn(self, inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:1231: in CALL_FUNCTION_EX
    self.call_function(fn, argsvars.items, kwargsvars)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:644: in call_function
    self.push(fn.call_function(self, args, kwargs))
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/lazy.py:90: in realize_and_forward
    return getattr(self.realize(), name)(*args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/nn_module.py:712: in call_function
    return variables.UserFunctionVariable(fn, source=source).call_function(
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:265: in call_function
    return super().call_function(tx, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:84: in call_function
    return tx.inline_user_function_return(
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:650: in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2233: in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2348: in inline_call_
    tracer.run()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:780: in run
    and self.step()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:743: in step
    getattr(self, inst.opname)(inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:462: in wrapper
    return inner_fn(self, inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:1231: in CALL_FUNCTION_EX
    self.call_function(fn, argsvars.items, kwargsvars)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:644: in call_function
    self.push(fn.call_function(self, args, kwargs))
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:311: in call_function
    return super().call_function(tx, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:265: in call_function
    return super().call_function(tx, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:84: in call_function
    return tx.inline_user_function_return(
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:650: in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2233: in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2348: in inline_call_
    tracer.run()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:780: in run
    and self.step()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:743: in step
    getattr(self, inst.opname)(inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:462: in wrapper
    return inner_fn(self, inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:1190: in CALL_FUNCTION
    self.call_function(fn, args, {})
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:644: in call_function
    self.push(fn.call_function(self, args, kwargs))
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:265: in call_function
    return super().call_function(tx, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:84: in call_function
    return tx.inline_user_function_return(
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:650: in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2233: in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2348: in inline_call_
    tracer.run()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:780: in run
    and self.step()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:743: in step
    getattr(self, inst.opname)(inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:462: in wrapper
    return inner_fn(self, inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:1190: in CALL_FUNCTION
    self.call_function(fn, args, {})
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:644: in call_function
    self.push(fn.call_function(self, args, kwargs))
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:265: in call_function
    return super().call_function(tx, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/functions.py:84: in call_function
    return tx.inline_user_function_return(
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:650: in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2233: in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2348: in inline_call_
    tracer.run()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:780: in run
    and self.step()
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:743: in step
    getattr(self, inst.opname)(inst)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:1301: in STORE_ATTR
    BuiltinVariable(setattr).call_function(
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py:650: in call_function
    result = handler(tx, *args, **kwargs)
../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py:1350: in call_setattr
    unimplemented(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

msg = 'setattr(UserDefinedObjectVariable) <function Module.__setattr__ at 0x7f56a8f51ea0>'

    def unimplemented(msg: str) -> NoReturn:
        assert msg != os.environ.get("BREAK", False)
>       raise Unsupported(msg)
E       torch._dynamo.exc.Unsupported: setattr(UserDefinedObjectVariable) <function Module.__setattr__ at 0x7f56a8f51ea0>
E       
E       from user code:
E          File "/home/adrian/.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/external_utils.py", line 25, in inner
E           return fn(*args, **kwargs)
E         File "/home/adrian/.conda/envs/lightning/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
E           return forward_call(*args, **kwargs)
E         File "/home/adrian/.conda/envs/lightning/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 833, in forward
E           args, kwargs = _root_pre_forward(self, self, args, kwargs)
E         File "/home/adrian/.conda/envs/lightning/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 543, in _root_pre_forward
E           _lazy_init(state, module)
E         File "/home/adrian/.conda/envs/lightning/lib/python3.10/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 162, in _lazy_init
E           state._is_root = True
E       
E       Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
E       
E       
E       You can suppress this exception and fall back to eager by setting:
E           import torch._dynamo
E           torch._dynamo.config.suppress_errors = True

../../.conda/envs/lightning/lib/python3.10/site-packages/torch/_dynamo/exc.py:193: Unsupported
```

**Comment by awaelchli** [comment]
Closing in favor of alternative approach #19280

**Pull request closed without merging**